File: Python/BOJ/12865 평범한 배낭.py
# import sys
# N, K = map(int, sys.stdin.readline().split())
# items = [[0,0]]+[list(map(int, sys.stdin.readline().split())) for _ in range(N)]
# dp = [[0]*(K+1) for _ in range(N+1)]
# for i in range(1, N+1):
#     for j in range(1, K+1):
#         if j < items[i][0]: # 아이템 무게보다 더 나가면
#             dp[i][j] = dp[i-1][j] # 이전 값
#         else : # 아니면
#             dp[i][j] = max(dp[i-1][j], dp[i-1][j - items[i][0]]+items[i][1]) # 이전 가치 / 현재가치 + 합할 수 있는 가치 합 비교
# print(dp[N][K])

import sys
N, K = map(int, sys.stdin.readline().split())
bag = {0 : 0}
for _ in range(N):
    weight, value = map(int, sys.stdin.readline().split())
    temp = {}
    for i, j in bag.items():
        w, v = weight + j, value + i
        if w < bag.get(v, K+1):
            temp[v] = w
    bag.update(temp)
print(max(bag.keys()))

File: Java/BOJ/12865 평범한 배낭.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        int N = Integer.parseInt(st.nextToken());
        int K = Integer.parseInt(st.nextToken());
        List<int[]> items = new ArrayList<>();
        for(int i = 0 ; i < N; i++) {
        	st = new StringTokenizer(br.readLine()," ");
        	int weight = Integer.parseInt(st.nextToken());
        	int value =  Integer.parseInt(st.nextToken());
        	items.add(new int[] {weight, value});
        }
        Collections.sort(items, (a,b) -> Integer.compare(b[1], a[1]));
        Map<Integer, Integer> bag = new HashMap<>();
        bag.put(0, 0);
        for(int[] item : items) {
        	int weight = item[0];
        	int value = item[1];
        	Map<Integer, Integer> temp = new HashMap<>(bag);
        	
        	for(Map.Entry<Integer, Integer> entry : bag.entrySet()) {
        		int i = entry.getKey();
        		int j = entry.getValue();
        		
        		int newValue = value + i;
        		int newWeight = weight + j;
        		
        		if(bag.getOrDefault(newValue, K+1) > newWeight){
        			temp.put(newValue, newWeight);
        		}
        	}
        	bag.putAll(temp);
        }
        int max = Collections.max(bag.keySet());
        System.out.println(max);
    }
}
